mypy/plugins/dataclasses.py

```diff
@@ -0,0 +1,279 @@
+    'dataclass',
+    'dataclasses.dataclass',
+}
+
+
+class DataclassAttribute:
+    def __init__(
+            self,
+            name: str,
+            is_in_init: bool,
+            has_default: bool,
+            line: int, column: int,
+    ) -> None:
+        self.name = name
+        self.is_in_init = is_in_init
+        self.has_default = has_default
+        self.line = line
+        self.column = column
+
+    def to_argument(self, info: TypeInfo) -> Argument:
+        return Argument(
+            variable=self.to_var(info),
+            type_annotation=info[self.name].type,
+            initializer=None,
+            kind=ARG_OPT if self.has_default else ARG_POS,
+        )
+
+    def to_var(self, info: TypeInfo) -> Var:
+        return Var(self.name, info[self.name].type)
+
+    def serialize(self) -> JsonDict:
+        return {
+            'name': self.name,
+            'is_in_init': self.is_in_init,
+            'has_default': self.has_default,
+            'line': self.line,
+            'column': self.column,
+        }
+
+    @classmethod
+    def deserialize(cls, data: JsonDict) -> "DataclassAttribute":
+        return cls(**data)
+
+
+class DataclassTransformer:
+    def __init__(self, ctx: ClassDefContext) -> None:
+        self._ctx = ctx
+
+    def transform(self) -> None:
+        ctx = self._ctx
+        info = self._ctx.cls.info
+        attributes = self.collect_attributes()
+        decorator_arguments = {
+            'init': _get_decorator_bool_argument(self._ctx, 'init', True),
+            'eq': _get_decorator_bool_argument(self._ctx, 'eq', True),
+            'order': _get_decorator_bool_argument(self._ctx, 'order', False),
+            'frozen': _get_decorator_bool_argument(self._ctx, 'frozen', False),
+        }
+
+        if decorator_arguments['init']:
+            _add_method(
+                ctx,
+                '__init__',
+                args=[attr.to_argument(info) for attr in attributes if attr.is_in_init],
+                return_type=NoneTyp(),
+            )
+            for stmt in self._ctx.cls.defs.body:
+                if isinstance(stmt, Decorator) and stmt.func.is_class:
+                    func_type = stmt.func.type
+                    if isinstance(func_type, CallableType):
+                        func_type.arg_types[0] = self._ctx.api.class_type(self._ctx.cls.info)
+
+        # Add an eq method, but only if the class doesn't already have one.
+        if decorator_arguments['eq'] and info.get('__eq__') is None:
+            cmp_tvar_def = TypeVarDef('T', 'T', 1, [], ctx.api.named_type('__builtins__.object'))
+            cmp_other_type = TypeVarType(cmp_tvar_def)
+            cmp_return_type = ctx.api.named_type('__builtins__.bool')
+
+            for method_name in ['__eq__', '__ne__']:
+                _add_method(
+                    ctx,
+                    method_name,
+                    args=[Argument(Var('other', cmp_other_type), cmp_other_type, None, ARG_POS)],
```

Comment:
I think it is dangerous to reuse the same type `cmp_other_type` for two different methods. (attrs plugin also does this, we can fix this separately).